src/main/java/org/opensearch/ad/transport/PMMLResultTransportAction.java

```diff
@@ -0,0 +1,137 @@
+    private static String DECISION_FUNCTION_FIELD = "decisionFunction";
+
+    @Inject
+    public PMMLResultTransportAction(
+        ActionFilters actionFilters,
+        TransportService transportService,
+        ADCircuitBreakerService adCircuitBreakerService,
+        NodeClient client
+    ) {
+        super(PMMLResultAction.NAME, transportService, actionFilters, PMMLResultRequest::new);
+        this.adCircuitBreakerService = adCircuitBreakerService;
+        this.client = client;
+    }
+
+    @Override
+    protected void doExecute(Task task, PMMLResultRequest request, ActionListener<PMMLResultResponse> listener) {
+
+        if (adCircuitBreakerService.isOpen()) {
+            listener.onFailure(new LimitExceededException(request.getAdID(), CommonErrorMessages.MEMORY_CIRCUIT_BROKEN_ERR_MSG));
+            return;
+        }
+
+        try {
+            MachineLearningClient mlClient = new MachineLearningNodeClient(client);
+            sendPredictionRequest(mlClient, request, listener);
+        } catch (Exception e) {
+            LOG.error(e);
+            listener.onFailure(e);
+        }
+    }
+
+    // utilize the ml client layer to make prediction calls
+    public void sendPredictionRequest(
+        MachineLearningClient mlClient,
+        PMMLResultRequest request,
+        ActionListener<PMMLResultResponse> listener
+    ) {
+        DataFrame inputData = loadDataFrame(request.getFeatureNames(), request.getFeatures());
+        mlClient.predict("pmml", null, inputData, request.getMlModelID(), ActionListener.wrap(response -> {
+            Map<String, Object> output = unloadDataFrame(response);
+            listener.onResponse(new PMMLResultResponse((boolean) output.get(OUTLIER_FIELD), (double) output.get(DECISION_FUNCTION_FIELD)));
+        }, listener::onFailure));
+    }
+
+    // load data frame from array of feature names and array of feature values
+    public DataFrame loadDataFrame(String[] featureNames, double[] features) throws AnomalyDetectionException {
+        if (featureNames.length != features.length) {
+            throw new AnomalyDetectionException("features names and features have different lengths");
+        }
+        List<Map<String, Object>> input = new ArrayList<>();
+        Map<String, Object> map = new LinkedHashMap<>();
+        for (int i = 0; i < features.length; i++) {
+            map.put(featureNames[i], features[i]);
+        }
+        input.add(map);
+        return DataFrameBuilder.load(input);
+    }
+
+    // unload data frame to a map of result names and result values
+    public Map<String, Object> unloadDataFrame(DataFrame results) throws AnomalyDetectionException {
+        if (results == null || results.size() == 0) {
+            throw new AnomalyDetectionException("null or empty response from the ML plugin");
+        }
+        Map<String, Object> result = new HashMap<>();
+        ColumnMeta[] header = results.columnMetas();
+        Row prediction = results.getRow(0);
```

Comment:
How about adding DataFrame example, so others can understand this parsing logic easily?

Reply:
Added in the comments!